torch/csrc/autograd/ir.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <Python.h>
+#include <memory>
+#include <vector>
+
+#include "torch/csrc/utils/object_ptr.h"
+
+namespace torch { namespace autograd {
+
+struct Output;
+struct Node;
+using output_list = std::vector<Output>;
+
+struct Output {
+  std::shared_ptr<Node> node;
```

Comment:
It seems that you have a reference cycle here. You have `Node` holding its `Output`s in a vector, and each output holds an owning pointer to the node.